sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/MultimapSideInput.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.fn.harness.state;
+
+import com.google.protobuf.ByteString;
+import java.io.IOException;
+import java.util.function.Supplier;
+import org.apache.beam.fn.harness.stream.DataStreams;
+import org.apache.beam.model.fnexecution.v1.BeamFnApi.StateRequest;
+import org.apache.beam.sdk.coders.Coder;
+
+/**
+ * An implementation of a multimap side input that utilizes the Beam Fn State API to fetch values.
+ *
+ * <p>TODO: Support block level caching and prefetch.
+ */
+public class MultimapSideInput<K, V> {
+  private final BeamFnStateClient beamFnStateClient;
```

Comment:
<!--thread_id:cc_145816474_t; commit:d1c0ffde59d0b17126deecb840cbeec36719d7a6; resolved:0-->
<!--section:context-quote-->
> **kennknowles** wrote:
> Got it

<!--section:body-->
Thought, considering it more, since it is not clear what fields are already set in that Supplier, it would probably be clearer to just pass the id and window to the constructor.

Reply:
This needs a much larger revamp to add a caching layer which will make the encoding/decoding aspect more complicated so I would rather leave this for now

Reply:
So the caching is based on a partially constructed proto message where you don't know what fields are set or unset?

Reply:
MultimapSideInput is only responsible for adding the encoded key that is being looked up, everything else has been populated by FnApiDoFnRunner.

Its like any other system which adds a piece of information that they are responsible for like HTTP headers, no single user knows all HTTP headers and only those that are interested read them.

Reply:
There's nothing in the API that ensures or indicates that the fields you are hoping to be set have been. At a minimum, `checkArgument` to be sure that the proto is in the state you expect. This isn't an open ended situation like HTTP headers.

Reply:
This is an intermediary and only passes forward the partial state request onto the next piece, why would it check the arguments for something it isn't responsible for?

Reply:
The interface for this class is `Iterable<V> get(K)`. So it seems like it is responsible for taking a `K` and giving back a `V`. So I'd expect the constructor to either (a) construct an object which will succeed at that or (b) let the caller know that it did not pass valid arguments to the constructor.

An alternative design that makes this all clear is passing the window and id to the constructor, and having this class build the whole proto message in `get`. This is probably just a better design, as the constructor can be reliable and the proto message can all be constructed in one place, by one class.

Reply:
(unresolving)